data/ipfs/ipfs.go

```diff
@@ -313,10 +329,29 @@ func (i *Handler) Retrieve(ctx context.Context, path string, maxSize int64) ([]b
 	// Save file to cache for future attempts
 	i.retrieveCache.Add(path, content)
 
-	log.Infow("retrieved file", "path", path, "size", fsize)
+	log.Infow("retrieved file", "path", path, "size", len(content))
 	return content, nil
 }
 
+func fetchFileContent(node files.Node) ([]byte, error) {
+	file := files.ToFile(node)
+	if file == nil {
+		return nil, fmt.Errorf("object is not a file")
+	}
+	defer file.Close()
+
+	fsize, err := file.Size()
+	if err != nil {
+		return nil, err
+	}
+
+	if fsize > MaxFileSizeBytes {
+		return nil, fmt.Errorf("file too big: %d", fsize)
+	}
+
+	return io.ReadAll(file)
```

Comment:
use `io.LimitReader` with `MaxFileSizeBytes` too; otherwise if `Size` is buggy you might still read gigabytes here.